### DISCUSSION TITLE
How to speed up getting all the solutions?

### DISCUSSION THREAD

**Discussion by ghost** [google/or-tools#3239] [CP-SAT questions] [answered]
Here I integate OR-Tools into my Java Program, and use CP-SAT Solver to get all the solution of Constraint Satisfaction Problem. But it is slow when there is a large amount of data, how to speed up the solving process?

**Comment by lperron** [comment, marked as answer]
create some partial, disjoint, assignment of variables, and search with workers in parallel.

**Reply by ghost** [reply]
But searching with workers in parallel can not be used to get all solutions in CP-SAT

**Comment by lperron** [comment]
no, multiples separate CP-SAT solvers, each working on a different split of
the search space.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by ghost** [reply]
Could you give me some simple and useful example codes? Thanks a lot.

**Comment by lperron** [comment]
enumerate all possible values of A.
Create one solver per value of A.
Fix A = value for each solver (1 different value per solver)
Run all solvers in parallel, everybody in its own thread.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by ghost** [reply]
Should I write all the parallel computing code? Do Google-OR Tools provide some Java API to create some parallel solvers for one big problem?

**Comment by lperron** [comment]
nothing complex, 1 thread per value of A.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by ghost** [reply]
Thanks! It really speed up a lot!!!! 
And here I have the final question, if i want to create a variable in discrete value such as [10,20,30,300]. How to represent it in CP-SAT Solver?

**Comment by lperron** [comment]
in python: var =
cp_model.NewIntVarFromDomain(cp_model.Domain.FromValues([10, 20, 30, 400]))
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by ghost** [reply]
Thanks!!! 
If the value of each A uses 1 thread, assuming that the range of A is [1,20000], can the system support the creation of 20000 threads?

**Comment by lperron** [comment]
most likely not. you can try.
You should be able to find libraries for thread pools.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by ghost** [reply]
Thanks! But how can I transfer the constraints of an existing model directly to another model without having to add them one by one? How can I modify  or copy exist models in Java?

**Reply by nthe** [reply]
There's a way to save/load the model, in the form of a protobuf object, to a file.
Take a look at examples over here - https://github.com/google/or-tools/blob/stable/ortools/sat/docs/model.md

**Reply by ghost** [reply]
It seems we can not save/load the model in Java?

**Comment by nthe** [comment]
Python threads are not truly parallel, and solving is a CPU bound job if I'm not mistaken.
You'll probably need to use process pool for that.

**Reply by ghost** [reply]
Thanks!
